exonum/src/events/noise/mod.rs

```diff
@@ -112,3 +114,190 @@ fn write_handshake_msg(
     Box::new(done(res.map_err(|e| e.into())))
 }
+
+pub fn read_handshake_msg(
+    input: &[u8],
+    noise: &mut NoiseWrapper,
+) -> Box<Future<Item = (usize, Vec<u8>), Error = io::Error>> {
+    let res = noise.read_handshake_msg(input);
+    Box::new(done(res.map_err(|e| e.into())))
+}
+
+#[cfg(test)]
+mod tests {
+    use futures::sync::{mpsc, mpsc::Receiver, mpsc::Sender};
+    use futures::{done, Future, Sink, Stream};
+    use tokio_core::net::{TcpListener, TcpStream};
+    use tokio_core::reactor::Core;
+    use tokio_io::AsyncRead;
+    use tokio_timer::{TimeoutStream, Timer};
+
+    use std::io;
+    use std::net::SocketAddr;
+    use std::thread;
+    use std::time::Duration;
+
+    use crypto::{gen_keypair_from_seed, Seed};
+    use events::codec::MessagesCodec;
+    use events::error::{into_other, log_error};
+    use events::noise::wrapper::{NoiseWrapper, NOISE_MAX_MESSAGE_LENGTH,
+                                 NOISE_MIN_HANDSHAKE_MESSAGE_LENGTH};
+    use events::noise::{read, read_handshake_msg, write, HandshakeParams, HandshakeResult,
+                        NoiseHandshake};
+
+    #[derive(Debug, PartialEq, Copy, Clone)]
+    pub enum HandshakeStep {
+        Default,
+        One(u8, usize),
+        Two(u8, usize),
+    }
+
+    const EMPTY_MESSAGE: usize = 0;
+    const SMALL_MESSAGE: usize = NOISE_MIN_HANDSHAKE_MESSAGE_LENGTH - 1;
+    const BIG_MESSAGE: usize = NOISE_MAX_MESSAGE_LENGTH + 1;
+
+    impl Default for HandshakeParams {
+        fn default() -> Self {
+            let (public_key, secret_key) = gen_keypair_from_seed(&Seed::new([0; 32]));
+            HandshakeParams {
+                max_message_len: 1024,
+                public_key,
+                secret_key,
+            }
+        }
+    }
+
+    #[test]
+    fn test_noise_normal_handshake() {
+        let addr: SocketAddr = "127.0.0.1:45001".parse().unwrap();
+        let addr2 = addr.clone();
+
+        let (sender, receiver) = mpsc::channel(0);
+        let receiver = add_timeout_millis(receiver, 500);
+        thread::spawn(move || run_handshake_listener(&addr2, sender));
+
+        // Use first handshake only to connect.
+        let _res = send_handshake(&addr, HandshakeStep::Default);
+        receiver.wait().next();
+
+        let res = send_handshake(&addr, HandshakeStep::Default);
+        println!("Send handshake result {:?}", res);
+        assert!(res.is_ok());
+    }
+
+    #[test]
+    fn test_noise_handshake_errors() {
+        let addr: SocketAddr = "127.0.0.1:45002".parse().unwrap();
+        let addr2 = addr.clone();
+
+        let (sender, receiver) = mpsc::channel(0);
+        let receiver = add_timeout_millis(receiver, 500);
+        thread::spawn(move || run_handshake_listener(&addr2, sender));
+
+        // Use first handshake only to connect.
+        let _res = send_handshake(&addr, HandshakeStep::Default);
+        receiver.wait().next();
+
+        let res = send_handshake(&addr, HandshakeStep::One(1, EMPTY_MESSAGE));
```

Comment:
Are you sure that multiple invocations of `send_handshake` won't mess up the other side of the channel? Resetting the channel after each handshake could be cleaner.